### PR TITLE
chore: migrate site to Next.js

### PR DESCRIPTION
**Pull request by Niznikr**
Migrate Paste website to Next.js for better DX and web vitals.

**Migration**

- Follow [migration guidelines](https://nextjs.org/docs/migrating/from-gatsby)
- `yarn start` -> `yarn dev`
- `yarn serve` -> `yarn start`
- Gatsby Link -> [Next Link](https://nextjs.org/docs/api-reference/next/link) (prop `to` -> `href`)
- Use [next/future/image](https://nextjs.org/blog/next-12-3#image-component) to be ready for Next 13 upgrade
- react-helmet -> [next/head](https://nextjs.org/docs/api-reference/next/head)
- [@next/mdx](https://nextjs.org/docs/advanced-features/using-mdx) with custom remark plugin to get mdx page headings
- Gatsby location -> [next/router](https://nextjs.org/docs/api-reference/next/router)
- Add google analytics [following example](https://github.com/vercel/next.js/tree/canary/examples/with-google-analytics)
- Fetch airtable data via [the js client](https://github.com/airtable/airtable.js/)
- Use a script to fetch navbar data and features from airtable to avoid api rate limits
- Gatsby useStaticQuery -> Next [getStaticProps](https://nextjs.org/docs/basic-features/data-fetching/get-static-props)
- Generate sitemap with [getServerSideProps](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props)
- Add site manifest
- Move gifs to same location as other images for blog posts
- Lazy load docsearch to minimize initial JS load
- Add Next ESLint plugin and [bundle analyzer](https://www.npmjs.com/package/@next/bundle-analyzer)
- Update env variables prefix `GATSBY_` -> `NEXT_PUBLIC_`
- Update docs

**Stats**

Gatsby - Local dev
`success Building development bundle - 187.168s` - cold cache
`success Building development bundle - 20.532s` - hot cache

Next.js - Local dev
`event - compiled client and server successfully in 2.9s (1074 modules)` - cold and hot cache

**Gatsby - preview deploy**

Screenshot 2023-02-03 at 10 56 02 AM

**Next.js - preview deploy**

Screenshot 2023-02-03 at 10 56 20 AM

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -225,11 +224,13 @@
...
     "typescript": "^4.9.4",
     "ui-box": "^5.4.0",
+    "unist-util-visit-esm": "npm:unist-util-visit@^4.1.2",
     "vite": "^4.0.1"
   },
   "resolutions": {
     "csstype": "3.0.11",
-    "playwright": "1.28.1"
+    "playwright": "1.28.1",
+    "@types/react": "^17.0.31"
```
Comment (shleewhite) [inline package.json:233]: Question: why is this resolution necessary?

**Review by shleewhite — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (Niznikr) [reply package.json:233]: The mdx library was allowing types v18 to get installed. Setting a resolution for now to avoid TS errors until we upgrade to 18.

File: .eslintignore [hunk 2]
```diff
@@ -1,6 +1,5 @@
 node_modules
 bower_components
-lib
```
Comment (TheSisb) [inline .eslintignore:3]: question: why this change?

**Review by SiTaggart — APPROVED** [review]
Pretty good test of our test coverage 🤞🏼

File: internal-docs/engineering/doc-site/airtable-integration.md [hunk 3]
```diff
@@ -9,30 +9,50 @@
 
 The overview tables and component/pattern headers all use an Airtable integration to gather the data displayed in each.
 
-The Airtable connection is done through the `gatsby-source-airtable` plugin. This uses an api key and pulls all of the Airtable information into a graphql schema. We then use this to populate the component or pattern statuses. Here is an example of the component overview page graphql query:
+The Airtable connection is done through the `Airtable JS client`. It uses an api key and allows us to query the Airtable information. We then use this to populate the component or pattern statuses. Here is an example of the query used to fetch all features:
```
Comment (TheSisb) [inline internal-docs/engineering/doc-site/airtable-integration.md:12]: praise: love these docs!

File: internal-docs/engineering/doc-site/airtable-integration.md [hunk 4]
```diff
@@ -9,30 +9,50 @@
...
+  })
+  .all();
+const items = features.map(({fields}) => fields);
 ```
 
-In the query, we’re filtering all of the Airtable data by status and Component_Category. Then getting the Feature, Documentation, Figma, etc table data.
+The data is then written to a JSON file and made available via the `utils/api` helper functions. The overview pages can then access this data using Next's `getStaticProps` and pass the information to other components:
+
+```js
+import {getAllComponents} from '../../utils/api';
```
Comment (TheSisb) [inline internal-docs/engineering/doc-site/airtable-integration.md:43]: praise: this approach is much more intuitive imo

File: internal-docs/engineering/doc-site/mdx.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# MDX
```
Comment (TheSisb) [inline internal-docs/engineering/doc-site/mdx.md:1]: praise: nice addition

File: packages/paste-codemods/package.json [hunk 6]
```diff
@@ -15,7 +15,7 @@
   "dependencies": {
     "chalk": "4.1.2",
     "execa": "3.2.0",
-    "globby": "10.0.1",
+    "globby": "npm:globby@10.0.1",
```
Comment (TheSisb) [inline packages/paste-codemods/package.json:18]: question: do we need the npm prefix here?

File: packages/paste-website/plugins/remark-headings-plugin.mjs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import {visit} from 'unist-util-visit-esm';
+import {toString} from 'mdast-util-to-string';
+
+// eslint-disable-next-line unicorn/consistent-function-scoping
+const transform = () => (tree) => {
+  const headings = [];
+
+  visit(tree, `heading`, (heading) => {
```
Comment (TheSisb) [inline packages/paste-website/plugins/remark-headings-plugin.mjs:8]: quibble: can we get some comments on the purpose of this plugin for future maintainers and how it's loaded.

File: packages/paste-website/scripts/fetch-data.mjs [hunk 8]
```diff
@@ -0,0 +1,115 @@
+import * as fs from 'fs/promises';
```
Comment (TheSisb) [inline packages/paste-website/scripts/fetch-data.mjs:1]: praise: I love this idea to prebuild this file for the website launch from the Airtable source data.

**Review by sandsng — APPROVED** [review]
This is cooooooooool

File: packages/paste-website/src/utils/DataUtils.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  'allPasteLayout',
+  'allPasteLibraries',
+  'allPastePattern',
+  'allPastePrimitive',
+  'allPasteThemePackage',
+]);
+
+const shouldFlatten = (key: string): boolean => KEYS_TO_FLATTEN.has(key);
+
+export const getNormalizedHeaderData = (data: ApiData): ApiData => {
```
Comment (TheSisb) [inline packages/paste-website/src/utils/DataUtils.ts:24]: quibble: can we add comments in this file for the exported functions for what their purpose is, and where they're used please?

**Review by TheSisb — COMMENTED** [review]
This is incredible work! Looks very well put together!

File: .eslintignore (hunk 2 above)
Comment (Niznikr) [reply .eslintignore:3]: Have a lib directory in the docs package now for holding things like google analytics (to match practices from the [next official examples](https://github.com/vercel/next.js/tree/canary/examples/with-google-analytics)).

File: packages/paste-codemods/package.json (hunk 6 above)
Comment (Niznikr) [reply packages/paste-codemods/package.json:18]: Yeah to avoid issues from `manypkg`
